android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Router.kt

```diff
@@ -4,111 +4,96 @@ import android.os.Bundle
-        get() = backStackManager.state.current.configuration
-
+    private lateinit var backStackFeature: BackStackFeature<C>
+    private lateinit var configurationFeature: ConfigurationFeature<C>
     lateinit var node: Node<V>
         internal set
 
     fun onAttach(savedInstanceState: Bundle?) {
         timeCapsule = AndroidTimeCapsule(savedInstanceState)
-        initConfigurationManager()
+        initFeatures()
     }
 
-    protected open val permanentParts: List<() -> Node<*>> =
-        emptyList()
-
-    private fun initConfigurationManager() {
-        backStackRibConnector = BackStackRibConnector(
-            permanentParts.map { it.invoke() },
-            this::resolveConfiguration,
-            NodeConnector.from(
-                node::attachChildNode,
-                node::attachChildView,
-                node::detachChildView,
-                node::detachChildNode
-            )
-        )
-
-        backStackManager = BackStackManager(
-            backStackRibConnector = backStackRibConnector,
+    private fun initFeatures() {
+        backStackFeature = BackStackFeature(
             initialConfiguration = initialConfiguration,
             timeCapsule = timeCapsule
         )
+
+        configurationFeature = ConfigurationFeature(
+            initialConfigurations = permanentParts,
+            timeCapsule = timeCapsule,
+            resolver = this::resolveConfiguration,
+            parentNode = node
+        )
+
+        binder.bind(backStackFeature.commands() to configurationFeature)
     }
 
     abstract fun resolveConfiguration(configuration: C): RoutingAction<V>
 
     fun onSaveInstanceState(outState: Bundle) {
-        backStackManager.accept(SaveInstanceState())
+        configurationFeature.accept(SaveInstanceState())
         timeCapsule.saveState(outState)
     }
 
     fun onLowMemory() {
-        backStackManager.accept(ShrinkToBundles())
+        // TODO add back support for this
     }
 
     fun onAttachView() {
-        backStackRibConnector.attachToView(backStackManager.state.backStack)
+        configurationFeature.accept(WakeUp())
     }
 
     fun onDetachView() {
-        backStackRibConnector.detachFromView(backStackManager.state.backStack)
+        configurationFeature.accept(Sleep())
     }
 
     fun onDetach() {
-        binder.clear()
-        backStackManager.dispose()
+        binder.dispose()
     }
 
-    fun replace(configuration: C) {
-        backStackManager.accept(Replace(configuration))
+    fun replace(configuration: Content) {
+        backStackFeature.accept(Replace(configuration))
     }
 
-    fun push(configuration: C) {
-        if (configuration is Overlay) {
-            backStackManager.accept(PushOverlay(configuration))
-        } else {
-            backStackManager.accept(Push(configuration))
-        }
+    fun push(configuration: Content) {
+        backStackFeature.accept(Push(configuration))
+    }
+
+    fun pushOverlay(configuration: Overlay) {
```

Comment:
I am not sure that dividing push into push/pushOverlay is really cool.
Of course it is more clean from library perspective, so now you can't push newRoot as an overlay, but from using perspective I would prefer to have one method.

Reply:
It's a trade-off for compile-time safety. You cannot accidentally put an overlay into content, or open a dialog in content-mode.